javalib/src/main/scala/java/nio/charset/UTF_8.scala

```diff
@@ -106,17 +106,29 @@ private[charset] object UTF_8 extends Charset("UTF-8", Array(
             if (length == -1) {
               finalize(CoderResult.malformedForLength(1))
             } else {
-              val decoded = {
-                @inline
-                def inArrayOr0(offset: Int): Int =
-                  if (inPos + offset < inEnd) inArray(inPos + offset)
-                  else 0 // 0 is not a valid next byte
-
-                val b2 = inArrayOr0(1)
-                if (length == 2) decode2(leading, b2)
-                else if (length == 3) decode3(leading, b2, inArrayOr0(2))
-                else decode4(leading, b2, inArrayOr0(2), inArrayOr0(3))
-              }
+              val decoded =
+                if (inPos + 1 >= inEnd)
+                  DecodedMultiByte(CoderResult.UNDERFLOW)
+                else {
+                  val b2 = inArray(inPos + 1)
+                  if (isInvalidNextByte(b2))
+                    DecodedMultiByte(CoderResult.malformedForLength(1))
+                  else if (length == 2)
+                    decode2(leading, b2)
+                  else if (inPos + 2 >= inEnd)
+                    DecodedMultiByte(CoderResult.UNDERFLOW)
+                  else {
+                    val b3 = inArray(inPos + 2)
+                    if (isInvalidNextByte(b3))
+                      DecodedMultiByte(CoderResult.malformedForLength(2))
+                    else if (length == 3)
+                      decode3(leading, b2, b3)
+                    else if (inPos + 3 >= inEnd)
+                      DecodedMultiByte(CoderResult.UNDERFLOW)
+                    else
+                      decode4(leading, b2, b3, inArray(inPos + 3))
```

Comment:
it might be nice to handle the last byte consistently with the rest

maybe then it's also a good idea to make decodeN assume correct continuation bytes so it doesn't need to do work validating that anymore

Reply:
I believe we can significantly increase the performance of the "good cases" by doing first a quick check for whether there are enough remaining bytes to have a chance of a correct result. Something like
```scala
val decoded = if (inPos + length <= inEnd) {
  // Enough bytes to decode one character
  if (length == 2)
    decode2(leading, inArray(inPos + 1))
  else if (length == 3)
    decode3(leading, inArray(inPos + 1), inArray(inPos + 2))
  else
    decode4(leading, inArray(inPos + 1), inArray(inPos + 2), inArray(inPos + 3))
} else {
  // Not enough bytes, either it's already known to be malformed, or it is an underflow
  var wellFormedUntilPos = inPos + 1
  while (wellFormedUntilPos < inEnd && !isInvalidNextByte(inArray(wellFormedUntilPos)))
    wellFormedUntilPos += 1
  if (wellFormedUntilPos == inEnd)
    DecodedMultiByte(CoderResult.UNDERFLOW)
  else
    DecodedMultiByte(CoderResult.malformedForLength(wellFormedUntilPos - inPos))
}
```
WDYT?

Reply:
This adds 1 range check to the end of the buffer halfway through a byte sequence (the initial range check + 1 for in the while per continuation byte), but you spare 1 range check for a 3 byte sequence and 2 range checks for a 4 bytes sequence that is not broken up.

My wild guess would be that for something like vietnamese text, which is relatively likely to have a 2 byte sequence at the end of an input buffer, but relatively unlikely to contain a 3 byte sequence, you variation would be slower, and for chinese, japanese or korean text, that are likely to consist mainly of 3 byte sequences, your variation would be faster, maybe significantly so.

Additionally, many of these checks are almost always the same for longer buffers (and who is going to be decoding in an 8 byte buffer?), so branch prediction could be doing a lot of work, to the point where I don't want to predict any performance characteristic.

Also, your code is a bit more compact (though it took me a bit more time to convince myself it was correct).

Do you want benchmarks for this? If so, are there any benchmarks examples I could follow? I wouldn't want to bet on any performance characteristic differences for japanese vs vienamese vs english text, either embedded in html or plain, and with various buffer sizes.

Reply:
Right. So there isn't a clear winner.

> Do you want benchmarks for this?

Nah, just pick whichever you prefer.